### PR TITLE
ssl_client2.c and ssl_server2.c extended to support TLS 1.3

### PR DESCRIPTION
**Pull request by hannestschofenig**
This removes the need to use dedicated example programs for TLS 1.3. Additionally some other bugs were corrected: the key derivation function incorrectly wrong length info, signature algorithm handling was corrected and extended, and pending alerts were previously not sent. Tested the basic functionality against Mint.

### CODE REVIEW

File: library/ssl_tls13_client.c [hunk 1]
```diff
@@ -792,15 +792,16 @@ int ssl_create_binder( mbedtls_ssl_context *ssl, unsigned char *psk, size_t psk_
     {
         ret = mbedtls_ssl_tls1_3_derive_secret( ssl, mbedtls_md_get_type( md ),
                             ssl->handshake->early_secret, hash_length,
-                            ( const unsigned char * )"res binder", sizeof( "res binder" ),
+                            ( const unsigned char * )"res binder", strlen ( "res binder" ),
```
Comment (hanno-becker) [inline library/ssl_tls13_client.c:795]: @hannestschofenig Why was this change made? I remember making exactly the reverse change in places recently. The length of a constant string can and should be evaluated at compile-time, IMO.

File: library/ssl_tls13_client.c [hunk 2]
```diff
@@ -3068,6 +3072,42 @@ static int ssl_server_hello_parse( mbedtls_ssl_context* ssl,
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
     }
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+    /* Configure signature algorithm. Compare the offered signature algorithm(s) 
+     * with the selected ciphersuite matches the signature algorithm.
```
Comment (hanno-becker) [inline library/ssl_tls13_client.c]: Minor: The grammar seems off here.

File: include/mbedtls/ssl.h [hunk 3]
```diff
@@ -1964,6 +1964,22 @@ void mbedtls_ssl_conf_verify( mbedtls_ssl_config *conf,
...
+/**
+ * \brief          Set the signature algorithms (Optional).
+ *
+ *                 If set, the signature algorithms will be advertised in 
+ *                 the signature_algorithms extension in the ClientHello of 
+ *                 TLS/DTLS 1.3. 
+ *
+ *
+ * \param conf     The SSL configuration to use.
+ * \param sig_algs A list of signature algorithms with the most preferred algorithm listed first.
```
Comment (hanno-becker) [inline include/mbedtls/ssl.h:1977]: The documentation should state how the end of the list is detected.

File: library/ssl_tls13_client.c [hunk 4]
```diff
@@ -2905,7 +2906,10 @@ static int ssl_server_hello_parse( mbedtls_ssl_context* ssl,
     unsigned int ext_size; /* size of an individual extension */
 
     const mbedtls_ssl_ciphersuite_t* suite_info; /* pointer to ciphersuite */
-
+    
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) && defined(MBEDTLS_X509_CRT_PARSE_C)
+    const int *alg;
```
Comment (hanno-becker) [inline library/ssl_tls13_client.c]: Minor: A short comment on the purpose of this field would be helpful.

File: library/ssl_tls13_client.c [hunk 5]
```diff
@@ -3068,6 +3072,42 @@ static int ssl_server_hello_parse( mbedtls_ssl_context* ssl,
...
+        else if( ( ssl->transform_negotiate->ciphersuite_info->mac == MBEDTLS_MD_SHA512 ) && 
+            ( *alg == MBEDTLS_SIG_ECDSA_SECP521r1_SHA512 ) ) 
+        {
+            ssl->handshake->signature_scheme = MBEDTLS_SIG_ECDSA_SECP521r1_SHA512; 
+            break;
+        }
+    }
+
+    /* Error if we are unable to configure an appropriate signature algorithm. */
+    if( ssl->conf->signature_schemes[0] == MBEDTLS_SIG_NONE)
```
Comment (hanno-becker) [inline library/ssl_tls13_client.c]: Minor: Space missing before `)`.

File: library/ssl_tls13_generic.c [hunk 6]
```diff
@@ -214,7 +214,7 @@ static int ssl_calc_finished_tls_sha256(
     /* create client finished_key */
     ret = mbedtls_ssl_tls1_3_hkdf_expand_label( MBEDTLS_MD_SHA256,
                           ssl->handshake->client_handshake_traffic_secret, 32,
-                          (const unsigned char*)"finished", sizeof( "finished" ),
+                          (const unsigned char*)"finished", strlen( "finished" ),
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c:217]: Same as above: Why are we using `strlen`?

File: library/ssl_tls13_generic.c [hunk 7]
```diff
@@ -847,7 +847,7 @@ int ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
     /*
      * Determine length of the signature scheme list
      */
-    for ( md = ssl->conf->signature_schemes; *md != SIGNATURE_NONE; md++ )
+    for ( md = ssl->conf->signature_schemes; *md != MBEDTLS_SIG_NONE; md++ )
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Minor and pre-existing: No space after `for`.

File: library/ssl_tls13_generic.c [hunk 8]
```diff
@@ -862,7 +862,7 @@ int ssl_write_signature_algorithms_ext( mbedtls_ssl_context *ssl,
      * Write signature schemes
      */
 
-    for ( md = ssl->conf->signature_schemes; *md != SIGNATURE_NONE; md++ )
+    for ( md = ssl->conf->signature_schemes; *md != MBEDTLS_SIG_NONE; md++ )
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Same here: No space after `for`.

File: library/ssl_tls13_generic.c [hunk 9]
```diff
@@ -896,6 +896,8 @@ int ssl_parse_signature_algorithms_ext( mbedtls_ssl_context *ssl,
     const unsigned char *end = buf + len;
     const int *md_cur;
     int offered_signature_scheme;
+    int got_common_sig_alg=0;
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Minor: Spaces missing before/after `=`.

File: library/ssl_tls13_generic.c [hunk 10]
```diff
@@ -2997,29 +3003,27 @@ static int ssl_read_certificate_postprocess( mbedtls_ssl_context* ssl )
...
+                mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL, MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE );
+                break;
+            case MBEDTLS_ERR_SSL_NO_CLIENT_CERTIFICATE:
+                mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL, MBEDTLS_SSL_ALERT_MSG_CERT_REQUIRED );
+                break;
+            case MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED:
+                mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL, MBEDTLS_SSL_ALERT_MSG_UNKNOWN_CA );
+                break;
+            default:
+                mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL, MBEDTLS_SSL_ALERT_MSG_UNSUPPORTED_CERT );
```
Comment (gbryant-arm) [inline library/ssl_tls13_generic.c:3022]: Coding style: skip line before and after curly brackets

File: library/ssl_tls13_keys.c [hunk 11]
```diff
@@ -373,27 +383,26 @@ int mbedtls_ssl_tls1_3_hkdf_expand_label(
...
+    // ----------------------------- DEBUG ---------------------------
 
-        mbedtls_printf( "Info ( %d ):", len );
-        for ( int i = 0; i < len; i++ )
-        {
-            mbedtls_printf( "%02x", info[i] );
-        }
-        mbedtls_printf( "\n" );
+    mbedtls_printf( "Info ( %d ):", len );
+    for ( int i = 0; i < len; i++ )
```
Comment (gbryant-arm) [inline library/ssl_tls13_keys.c]: Coding style: superfluous space after `for`

**Comment by gbryant-arm** [comment]
The examples are still using `mbedtls_ssl_conf_export_keys_ext_cb()` and `mbedtls_ssl_tls_prf()` if `MBEDTLS_SSL_EXPORT_KEYS` is set, but we can fix that in a separate PR (https://github.com/hannestschofenig/mbedtls/pull/25)

File: library/ssl_tls13_server.c [hunk 12]
```diff
@@ -4533,89 +4571,43 @@ int mbedtls_ssl_handshake_server_step( mbedtls_ssl_context *ssl )
...
-            {
-                MBEDTLS_SSL_DEBUG_RET( 1, "ssl_read_certificate_process", ret );
-                return( ret );
-            }
             break;
 
             /* ----- READ CLIENT CERTIFICATE VERIFY ----*/
 
         case MBEDTLS_SSL_CLIENT_CERTIFICATE_VERIFY:
             ret=ssl_read_certificate_verify_process( ssl );
```
Comment (gbryant-arm) [inline library/ssl_tls13_server.c]: Minor: Space missing around `=`

**Pull request closed without merging**